msm/src/serialization/witness.rs

```diff
@@ -112,6 +140,44 @@ impl<Fp: PrimeField> Env<Fp> {
             }
         }
     }
+
+    pub fn add_rangecheck4_table_value(&mut self, i: usize) {
+        if i < (1 << 4) {
+            self.lookup_t_multiplicities_rangecheck4[i] += Fp::one();
+        } else {
+            self.lookup_t_multiplicities_rangecheck4[0] += Fp::one();
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.rangecheck15_lookups = vec![];
+        self.rangecheck4_lookups = vec![];
+    }
+
+    pub fn get_rangecheck4_normalized_multipliticies(
+        &self,
+        domain: EvaluationDomains<Fp>,
+    ) -> Vec<Fp> {
+        let mut m = vec![Fp::zero(); 1 << 4];
+        self.lookup_multiplicities_rangecheck4
+            .into_iter()
+            .zip(self.lookup_t_multiplicities_rangecheck4.iter())
+            .enumerate()
+            .for_each(|(i, (m_f, m_t))| m[i] = m_f / m_t);
```

Comment:
Why do we need to do this (field? or is it integer?) division here?  Is it related to the dummy values?..

Reply:
See equation 14: https://eprint.iacr.org/2022/1530.pdf